File: apps/ios/Reels/Features/Picker/PickerViewModel.swift
import Foundation
import Combine

/// ViewModel for the Picker feature.
@MainActor
final class PickerViewModel: ObservableObject {
    @Published var plans: [PickerPlanSummary] = []
    @Published var currentPlan: PickerPlanDetail?
    @Published var filmSearchResults: [FilmSearchResult] = []
    @Published var showtimes: [PickerShowtime] = []
    @Published var isLoading = false
    @Published var error: String?
    @Published var shareUrl: String?
    @Published var createdPlanId: String?

    private var searchTask: Task<Void, Never>?

    // MARK: - Film Search (with 300ms debounce)

    func searchFilms(query: String) {
        searchTask?.cancel()
        guard !query.isEmpty else {
            filmSearchResults = []
            return
        }

        searchTask = Task {
            try? await Task.sleep(nanoseconds: 300_000_000)
            guard !Task.isCancelled else { return }

            do {
                let response = try await PickerAPI.searchFilms(query: query)
                filmSearchResults = response.results
            } catch {
                if !Task.isCancelled {
                    self.error = error.localizedDescription
                }
            }
        }
    }

    // MARK: - Showtimes

    func loadShowtimes(filmTitle: String, city: String? = nil, cinema: String? = nil, date: String? = nil) async {
        isLoading = true
        error = nil
        do {
            let response = try await PickerAPI.getShowtimes(filmTitle: filmTitle, city: city, cinema: cinema, date: date)
            showtimes = response.showtimes
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    // MARK: - Create Plan

    func createPlan(input: CreatePlanInput) async {
        isLoading = true
        error = nil
        do {
            let response = try await PickerAPI.createPlan(input: input)
            shareUrl = response.shareUrl
            createdPlanId = response.planId
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    // MARK: - Load Plan

    func loadPlan(planId: String) async {
        isLoading = true
        error = nil
        do {
            currentPlan = try await PickerAPI.getPlan(planId: planId)
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    // MARK: - Join Plan

    func joinPlan(planId: String, displayName: String, guestToken: String?) async -> JoinPlanResponse? {
        do {
            let response = try await PickerAPI.joinPlan(input: JoinPlanInput(planId: planId, displayName: displayName, guestSessionToken: guestToken))
            if let token = response.sessionToken {
                UserDefaults.standard.set(token, forKey: "picker-guest-\(planId)")
            }
            await loadPlan(planId: planId)
            return response
        } catch {
            self.error = error.localizedDescription
            return nil
        }
    }

    // MARK: - Vote

    func vote(participantId: String, showtimeId: String, status: String) async {
        do {
            let input = VoteInput(participantId: participantId, votes: [.init(showtimeId: showtimeId, status: status)])
            _ = try await PickerAPI.vote(input: input)
        } catch {
            self.error = error.localizedDescription
        }
    }

    // MARK: - Confirm

    func confirmShowtime(planId: String, showtimeId: String) async {
        isLoading = true
        do {
            _ = try await PickerAPI.confirmPlan(input: ConfirmInput(planId: planId, showtimeId: showtimeId))
            await loadPlan(planId: planId)
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    // MARK: - My Plans

    func loadMyPlans() async {
        do {
            let response = try await PickerAPI.myPlans()
            plans = response.plans
        } catch {
            self.error = error.localizedDescription
        }
    }
}

File: apps/android/app/src/main/java/com/reels/app/features/picker/PickerViewModel.kt
package com.reels.app.features.picker

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.launch

data class PickerUiState(
    val myPlans: List<PickerPlanSummary> = emptyList(),
    val searchResults: List<FilmSearchResult> = emptyList(),
    val showtimes: List<PickerShowtime> = emptyList(),
    val currentPlan: PickerPlanDetail? = null,
    val isLoading: Boolean = false,
    val isSearching: Boolean = false,
    val error: String? = null,
    val shareUrl: String? = null,
    val justCreatedPlanId: String? = null,
)

@OptIn(FlowPreview::class)
class PickerViewModel(private val repository: PickerRepository) : ViewModel() {

    private val _uiState = MutableStateFlow(PickerUiState())
    val uiState: StateFlow<PickerUiState> = _uiState.asStateFlow()

    private val _searchQuery = MutableStateFlow("")

    init {
        viewModelScope.launch {
            _searchQuery
                .debounce(300)
                .distinctUntilChanged()
                .filter { it.length >= 2 }
                .collect { query -> performSearch(query) }
        }
    }

    fun loadMyPlans() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            try {
                val response = repository.myPlans()
                _uiState.value = _uiState.value.copy(
                    myPlans = response.plans,
                    isLoading = false,
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Failed to load plans",
                )
            }
        }
    }

    fun updateSearchQuery(query: String) {
        _searchQuery.value = query
        if (query.length < 2) {
            _uiState.value = _uiState.value.copy(searchResults = emptyList())
        }
    }

    private suspend fun performSearch(query: String) {
        _uiState.value = _uiState.value.copy(isSearching = true)
        try {
            val response = repository.searchFilms(query)
            _uiState.value = _uiState.value.copy(
                searchResults = response.results,
                isSearching = false,
            )
        } catch (e: Exception) {
            _uiState.value = _uiState.value.copy(
                isSearching = false,
                error = e.message,
            )
        }
    }

    fun loadShowtimes(filmTitle: String, city: String?) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            try {
                val response = repository.getShowtimes(filmTitle, city)
                _uiState.value = _uiState.value.copy(
                    showtimes = response.showtimes,
                    isLoading = false,
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message,
                )
            }
        }
    }

    fun createPlan(request: CreatePlanRequest) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            try {
                val response = repository.createPlan(request)
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    shareUrl = response.shareUrl,
                    justCreatedPlanId = response.planId,
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message,
                )
            }
        }
    }

    fun loadPlan(planId: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            try {
                val plan = repository.getPlan(planId)
                _uiState.value = _uiState.value.copy(
                    currentPlan = plan,
                    isLoading = false,
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message,
                )
            }
        }
    }

    fun joinPlan(planId: String, displayName: String, guestToken: String?) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            try {
                repository.joinPlan(
                    JoinPlanRequest(
                        planId = planId,
                        displayName = displayName,
                        guestSessionToken = guestToken,
                    ),
                )
                loadPlan(planId)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message,
                )
            }
        }
    }

    fun vote(participantId: String, votes: List<VoteRequest.VoteEntry>) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(error = null)
            try {
                repository.vote(VoteRequest(participantId = participantId, votes = votes))
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.message)
            }
        }
    }

    fun confirmPlan(planId: String, showtimeId: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            try {
                repository.confirmPlan(ConfirmRequest(planId = planId, showtimeId = showtimeId))
                loadPlan(planId)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message,
                )
            }
        }
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }

    fun clearCreatedPlan() {
        _uiState.value = _uiState.value.copy(justCreatedPlanId = null, shareUrl = null)
    }
}
